<a href="https://colab.research.google.com/github/rajnish80130/feature-extraction-PCA/blob/main/Task_PCA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### `Task` How dimensionality reduction using Principal Component Analysis (PCA) on the Wine Quality dataset contributes to improving the classification accuracy and efficiency of wine type.

Note : Use KNN for Classification.

Data Link :  [Wine Data](https://docs.google.com/spreadsheets/d/e/2PACX-1vQDVwxneOKOaJL13QMhkAhYrgWlH1tICY7RacUnj_lL8m9uUWaaUf3p7bScNyh_D2Rvt7nc1q11adSy/pub?gid=647503637&single=true&output=csv)

In [19]:
# Data Loading
import pandas as pd
import numpy as np
wine_data_path = "https://docs.google.com/spreadsheets/d/e/2PACX-1vQDVwxneOKOaJL13QMhkAhYrgWlH1tICY7RacUnj_lL8m9uUWaaUf3p7bScNyh_D2Rvt7nc1q11adSy/pub?gid=647503637&single=true&output=csv"
df = pd.read_csv(wine_data_path)

In [20]:
# Your Code goes Here
df.head()

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   
1  white            6.3              0.30         0.34             1.6   
2  white            8.1              0.28         0.40             6.9   
3  white            7.2              0.23         0.32             8.5   
4  white            7.2              0.23         0.32             8.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   

   sulphates  alcohol  quality  
0       0.45      8.8        6  
1       0.49      9.5        6  
2       0.44     10.1        6  
3       0.40      9.9        6  
4       0.40      9.9        6

In [21]:
df.shape

(6497, 13)

In [25]:
np.sum(df.isna())

type                     0
fixed acidity           10
volatile acidity         8
citric acid              3
residual sugar           2
chlorides                2
free sulfur dioxide      0
total sulfur dioxide     0
density                  0
pH                       9
sulphates                4
alcohol                  0
quality                  0
dtype: int64

In [26]:
df = df.dropna()

In [27]:
np.sum(df.isna())

type                    0
fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [28]:
df.shape

(6463, 13)

In [29]:
X = df.iloc[:,1:]
y = df.iloc[:,0]

In [36]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y, test_size = 0.2, random_state=7)

In [31]:
print(X_train.shape)
print(X_test.shape)

(5170, 12)
(1293, 12)


In [32]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [33]:
knn = KNeighborsClassifier(n_neighbors = 10)

knn.fit(X_train,y_train)
y_pred = knn.predict(X_test)

accuracy_score(y_pred,y_test)

0.9899458623356535

In [54]:
#PCA - step 1 (Scale the data)
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [55]:
from sklearn.decomposition import PCA

pca = PCA(n_components = 3)

X_train_trf = pca.fit_transform(X_train)
X_test_trf = pca.transform(X_test)

In [56]:
X_train_trf.shape

(5170, 3)

In [57]:
knn = KNeighborsClassifier(n_neighbors = 10)

knn.fit(X_train_trf,y_train)
y_pred_trf = knn.predict(X_test_trf)

accuracy_score(y_pred_trf,y_test)

0.9806651198762568

In [58]:
for i in range(1,13):
  pca = PCA(n_components = i)

  X_train_trf = pca.fit_transform(X_train)
  X_test_trf = pca.transform(X_test)

  knn = KNeighborsClassifier(n_neighbors = 5)

  knn.fit(X_train_trf,y_train)
  y_pred_trf = knn.predict(X_test_trf)

  print(accuracy_score(y_pred_trf,y_test))


0.9822119102861562
0.9822119102861562
0.9791183294663574
0.988399071925754
0.9891724671307038
0.9891724671307038
0.9891724671307038
0.9930394431554525
0.9907192575406032
0.991492652745553
0.991492652745553
0.9922660479505027


In [59]:
pca.explained_variance_ratio_

array([0.25282683, 0.220943  , 0.13764084, 0.08789284, 0.07102566,
       0.05488029, 0.0476827 , 0.04228503, 0.03844969, 0.0247637 ,
       0.01883195, 0.00277748])

In [60]:
np.cumsum(pca.explained_variance_ratio_)

array([0.25282683, 0.47376983, 0.61141066, 0.69930351, 0.77032916,
       0.82520946, 0.87289215, 0.91517719, 0.95362688, 0.97839057,
       0.99722252, 1.        ])

## without pca we achieve 98% accuracy but after pca we can achieve upto 99%